cabal-install/Distribution/Client/Init/FileCreators.hs

```diff
@@ -327,241 +339,277 @@ findNewName oldName = findNewName' 0
-           _               -> empty
+  , fieldP "extra-source-files" (maybeToFlag $ (formatExtraSourceFiles <$> extraSrc c))
+           ["Extra files to be distributed with the package, such as examples or a README."]
+           True
+  ]
+  ++
+  (case packageType c of
+     Flag Executable -> [executableStanza]
+     Flag Library    -> [libraryStanza]
+     Flag LibraryAndExecutable -> [libraryStanza, executableStanza]
+     _               -> [])
+  ++
+  if eligibleForTestSuite c then [testSuiteStanza] else []
 
-       , if eligibleForTestSuite c then testSuiteStanza else empty
-       ]
  where
    specVer = fromMaybe defaultCabalVersion $ flagToMaybe (cabalVersion c)
 
    licenseStr | specVer < CabalSpecV2_2 = prettyShow . licenseFromSPDX <$> license c
               | otherwise               = prettyShow                   <$> license c
 
-   generateBuildInfo :: BuildType -> InitFlags -> Doc
-   generateBuildInfo buildType c' = vcat
-     [ fieldS "other-modules" (listField otherMods)
-              (Just $ case buildType of
+   generateBuildInfo :: BuildType -> InitFlags -> [PrettyField [String]]
+   generateBuildInfo buildType c' = catMaybes
+     [ fieldP "other-modules" (formatOtherModules <$> maybeToFlag otherMods)
+       [ case buildType of
                  LibBuild    -> "Modules included in this library but not exported."
-                 ExecBuild -> "Modules included in this executable, other than Main.")
-              True
-
-     , fieldS "other-extensions" (listField (otherExts c'))
-              (Just "LANGUAGE extensions used by modules in this package.")
-              True
-
-     , fieldS "build-depends" ((++ myLibDep) <$> listField (dependencies c'))
-              (Just "Other library packages from which modules are imported.")
-              True
-
-     , fieldS "hs-source-dirs" (listFieldS (case buildType of
-                                            LibBuild  -> sourceDirs c'
-                                            ExecBuild -> applicationDirs c'))
-              (Just "Directories containing source files.")
-              True
-
-     , fieldS "build-tools" (listFieldS (buildTools c'))
-              (Just "Extra tools (e.g. alex, hsc2hs, ...) needed to build the source.")
-              False
-
-     , field  "default-language" (language c')
-              (Just "Base language which the package is written in.")
-              True
+                 ExecBuild -> "Modules included in this executable, other than Main."]
+       True
+
+     , fieldP "other-extensions" (maybeToFlag (formatOtherExtensions <$> otherExts c))
+       ["LANGUAGE extensions used by modules in this package."]
+       True
+
+     , fieldP "build-depends" (maybeToFlag (formatDependencyList <$> buildDependencies))
+       ["Other library packages from which modules are imported."]
+       True
+
+     , fieldP "hs-source-dirs"
+       (maybeToFlag (formatHsSourceDirs <$> case buildType of
```

Comment:
I'd make `fieldPAla` variant which takes `Newtype a b =>`. and `(a -> b)` "pack" function.

Yet, I wonder what we should do so we don't need to copy `formatHsSourceDirs`. The `hs-source-dirs`  + `formatHsSourceDirs` coupling is already done in `FieldGrammar` module, we shouldn't need to copy & paste that.

Reply:
Yeah, I was thinking about that as well. The problem with the `FieldGrammar` module is it also ties it to where the field exists in the large structure (the `ALens' s a'` part of most of the typeclass functions). I'm wondering if we should re-package the field name and (optional) newtype pack into some structure that could then be shared, that's definitely a more invasive change though.

I went with the approach here of exporting the `formatXyz` (which as pointed out isn't exactly accurate because you could also use it to `unpackXyz`) because it was the least invasive and kept the newtype packing and formatting as a unit without having to expose the implementation details (aka the newtype wrapper) of any given "formatting". We could instead of exporting `formatHsSourceDirs` export the `FSep FilePathNT` part of it (as you mentioned), but then you lose the semantic level of "this is how you format the hs-source-dirs field" and instead you leak the low level implementation which is "you pack them using `FSep`.

Also, if at some point in the future you wanted to change how hs-source-dirs was formatted you wouldn't need to change anything outside of `FieldGrammar.hs` in the current form. Conversely, if you used the `fieldPAla` approach then you'd need to know that when you change the `FSep` newtype wrapper in `FieldGrammar.hs` then you also need to change `fieldPAla` in `Init/FileCreators.hs` which is bound to be forgotten.

Reply:
I was thinking more of type like:

```haskell
fieldP
   :: (Pretty b, Newtype b a)
   => FieldName
   -> Flag a
   -> Doc
```

which would be used as

```diff
-   , fieldP "hs-source-dirs"
-       (maybeToFlag (formatHsSourceDirs <$> case buildType of
+   , fieldPAla "hs-source-dirs" formatHsSourceDirs
+       (maybeToFlag (case buildType of
```

Reply:
The above decouples "what" from "how it looks", when how it looks doesn't come from the type directly.

Reply:
Ahh gotcha, yeah that makes sense and is much cleaner.